docs/elastic.rst

```diff
@@ -0,0 +1,358 @@
+    tf.keras.backend.set_session(tf.Session(config=config))
+
+    dataset = ...
+    model = ...
+
+    opt = keras.optimizers.Adadelta(lr * hvd.size())
+    opt = hvd.DistributedOptimizer(opt)
+
+    model.compile(loss=keras.losses.sparse_categorical_crossentropy,
+                  optimizer=opt,
+                  metrics=['accuracy'])
+
+    def on_state_reset():
+        tf.keras.backend.set_value(model.optimizer.lr, lr * hvd.size())
+
+    state = hvd.elastic.KerasState(model, batch=100, epoch=0)
+    state.register_reset_callbacks([on_state_reset])
+
+    callbacks = [
+        hvd.elastic.CommitStateCallback(state),
+        hvd.elastic.UpdateBatchStateCallback(state),
+        hvd.elastic.UpdateEpochStateCallback(state),
+    ]
+
+    if hvd.rank() == 0:
+        callbacks.append(keras.callbacks.ModelCheckpoint('./checkpoint-{epoch}.h5'))
+
+    @hvd.elastic.run
+    def train(state):
+        model.fit(dataset,
+                  steps_per_epoch=500 // hvd.size(),
+                  callbacks=callbacks,
+                  epochs=epochs - state.epoch,
+                  verbose=1 if hvd.rank() == 0 else 0)
+
+    train(state)
+
+
+Elastic PyTorch
+~~~~~~~~~~~~~~~
+
+.. code-block:: python
+    :emphasize-lines: 14,15,28,31,36,37
+
+    import torch
+    import horovod.torch as hvd
+
+    hvd.init()
+
+    torch.cuda.set_device(hvd.local_rank())
+
+    dataset = ...
+    model = ...
+
+    optimizer = optim.SGD(model.parameters(), lr * hvd.size())
+    optimizer = hvd.DistributedOptimizer(optimizer)
+
+    @hvd.elastic.run
+    def train(state):
+        batch_offset = state.batch
+        for state.epoch in range(state.epoch, epochs):
+            for state.batch in range(state.batch, batches_per_epoch):
+                data, target = get_random_batch()
+
+                optimizer.zero_grad()
+                output = model(data)
+                loss = F.nll_loss(output, target)
+                loss.backward()
+                optimizer.step()
+
+                if state.batch % batches_per_commit == 0:
+                    state.commit()
+            state.batch = 0
+
+    def on_state_reset():
+        # adjust learning rate on reset
+        for param_group in optimizer.param_groups:
+            param_group['lr'] = lr * hvd.size()
+
+    state = hvd.elastic.TorchState(model, optimizer, batch=0, epoch=0)
+    state.register_reset_callbacks([on_state_reset])
+    train(state)
+
+
+Running with horovodrun
+~~~~~~~~~~~~~~~~~~~~~~~
+
+Elastic training jobs are started using the ``horovodrun`` command line tool. The major difference when launching
+elastic jobs is that hosts are not specified explicitly, but instead **discovered** at runtime.  The most general way
+to allow Horovod to discover available hosts is to provide a ``--host-discovery-script`` when launching the job:
+
+.. code-block:: bash
+
+    $ horovodrun -np 8 --host-discovery-script discover_hosts.sh python train.py
```

Comment:
Will this `discover_hosts.sh` script be reused after `HorovodInternalError` or `HostsUpdatedInterrupt` is raised?

Reply:
Hey @zw0610, yes, the `discover_hosts.sh` is used by the driver, while `HorovodInternalError` and `HostsUpdatedInterrupt` are raised on the workers.  On the driver side, it will continue calling `discover_hosts.sh` forever to update the set of known hosts.  On the worker side, if one of those exceptions is raised, a reset event will happen, where the results of the `discover_hosts.sh` script will be used to add or remove workers from the group.

Reply:
I see. Thanks a lot :)